File: 146. LRU Cache/sol.py
class Node:
    def __init__(self, key: int, value: int):
        self.key = key
        self.value = value
        self.prev = self.next = None
        
class LRUCache:

    def __init__(self, capacity: int):
        self.cache = {}
        
        self.left = Node(0, 0)
        self.right = Node(0, 0)
        self.left.next = self.right
        self.right.prev = self.left

        self.capacity = capacity
        self.size = 0
        
    def remove(self, node: Node) -> None:
        print("  remove :", node.key, node.value)
        prev = node.prev
        next = node.next

        prev.next = next
        next.prev = prev


    
    def insert(self, node: Node) -> None:
        print("  insert :", node.key, node.value)
        prev = self.right.prev
        prev.next = node
        self.right.prev = node
        node.prev = prev
        node.next = self.right



    def get(self, key: int) -> int:
        print("get :", key)
        if key in self.cache:
            self.remove(self.cache[key])
            self.insert(self.cache[key])
            return self.cache[key].value
        return -1

    def put(self, key: int, value: int) -> None:
        print("put :", key, value)
        if key in self.cache:
            self.remove(self.cache[key])
            self.insert(self.cache[key])
            self.cache[key].value = value
            return 

        if self.size < self.capacity:
            self.size += 1
        else:
            first = self.left.next
            print("   removing first : ", first.key, first.value)
            self.remove(first)
            del self.cache[first.key]

        node = Node(key, value)
        self.insert(node)
        self.cache[key] = node


    



        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 146. LRU Cache/sol.go
package LRUCache

type node struct {
	key   int
	value int
	prev  *node
	next  *node
}

type LRUCache struct {
	capacity int
	size     int
	cache    map[int]*node
	left     *node
	right    *node
}

func Constructor(capacity int) LRUCache {

	lruCache := LRUCache{
		capacity: capacity,
		size:     0,
		cache:    make(map[int]*node),
	}

	lruCache.left = &node{0, 0, nil, nil}
	lruCache.right = &node{0, 0, nil, nil}
	lruCache.left.next = lruCache.right
	lruCache.right.prev = lruCache.left

	return lruCache
}

func (this *LRUCache) remove(n *node) {
	prev := n.prev
	next := n.next
	prev.next = next
	next.prev = prev
	delete(this.cache, n.key)
}

func (this *LRUCache) insert(n *node) {
	prev := this.right.prev
	prev.next = n
	n.prev = prev
	n.next = this.right
	this.right.prev = n
	this.cache[n.key] = n
}

func (this *LRUCache) Get(key int) int {
	n, exists := this.cache[key]

	if !exists {
		return -1
	}

	this.remove(n)
	this.insert(n)
	return n.value

}

func (this *LRUCache) Put(key int, value int) {
	n, exists := this.cache[key]
	if exists {
		this.remove(n)
		this.insert(n)
		n.value = value
		return
	}

	if this.size < this.capacity {
		this.size++
	} else {
		first := this.left.next
		this.remove(first)
	}
	n = &node{key, value, nil, nil}
	this.insert(n)

}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */
